### ISSUE TITLE
Checklist for 4.0.0

### ISSUE THREAD

**Issue by JohN100x1** [JohN100x1/WOTR_IsekaiMod#71]
## Implementations
- [x] Fix `Omega Strike` to apply to touch attacks.
- [x] Add Class variant features. (#69)
- [x] Add Mod support for modded spells (#72)
- [x] Fix "stuck on loading screen" bug
- [x] Review all changes

## Release Tasks
- [x] Copy localization
- [x] Update CHANGELOG.md
- [x] Update README.md
- [x] Update Info.json
- [x] Update Repository.json

## Post-Release Tasks
- Update mod description
- Update mod version
- New screenshots
- Release description

**Comment by kjk001** [comment]
yeah I really have to get started on that localization so we don't spam all the users log with the warning that Strings are automatically recreated, don't I?

Meanwhile preliminary Changelog and still open ToDo List:
 - internal logic switch to directly use TTCCore where possible rather than code "inspired" from them 
 - merge spellbook logic to merge the spellbooks of the major classes into the isekai book rather than add each spell manually
 - patch for some other mod that turned deity selection into a recursive selection call onto itself which made it hard to select deities
 - Blessing of the Mythic Overwhelming Feature (grants abilities of a mythic class besides your own, possible selections include spells, trickster, Azata, Angel, Lich Feats, only Azata and Tickster Feats are what I would call well tested at this point though)
 - Major Class Progression rework, thief feats removed from all archetypes that had them, instead players can now select a base class from which their Isekai Hero Class descends, current choices are:
1.  Barbarian Legacy - Ball of Rage, mostly grants Rage
2. Bard Legacy - Musical Prodige, for the hero that knows  they way to his harems heart is through a good song
3. Dread Knight Legacy - Dread Lord, only available with Expanded Content installed, the anti paladin
4.  Paladin Legacy - Hero of Light, smite your enemies with the power of the light
5. Kineticist Legacy - Kinetic Knight, melee fighter Kineticist (while not required I advice installing Expandended Element as their energize weapon feat is what really makes this archetype even work)
6. Kineticist Legacy - Kinetic Lord, the default Kineticist
7. Magus Legacy - Spellblade, **Attention Still requires any form of descent testing, so only use it if you are willing to essentially be the betatester for the Legacy at the moment**
8. Oracle Legacy - Seeker of Truth, choose a curse, mystery, or revelation every few levels
9. Rogue Legacy - Supernatural Thief, all the talents I removed from the default template
10. Shaman Legacy - Spirit Beacons, hex hex grant me power my spirit
11. Sorcerer Legacy - Chimera, grants a bloodline or bloodline feat every few levels
12. Tactician Legacy - Isekai Tactician, not a baseclass by itself but rather a mixture of several like hunter and Monster Tactician to create a path focused around tactical feats

Paladin and Dread Knight Legacy are appropriatly Alignment Restricted.
**ToDo** restrict the two Kinetic Legacies to ensure you can only ever pick one, please do not try to combine these.
**ToDo** the archetypes might need to be balanced against each other, because being able to pick a bloodline or Mystery every few levels is truly worthy of a Cheater *coughs* Sorry I meant Protagonist ;) while Rogue and Paladin variants actually are less than their base class in terms of benefits.

Isekai Protagonist has access to all variants for their base feature.
Hero only has access to heroic ones like Paladin...
Edge Lord is limited to melee focused Legacies like the Kinetic Knight, Rogue, Barbarian...
Villain obviously does not get choices like Barbarian(because you are meant to be a planner not a brute that just hits everything), Paladin (this one should explain itself) or Bard(you are not a Disney Villain!)...
God Emperor does not get a default selection because it was missing Rogue Talents in their original Balancing against each other.

Every Version can Dualclass by picking the Overwhelming Ability - Legacy Class Feature, which currently imposes no restrictions on what archetype you pick so even the God Emperor can still get access to a legacy.
**ToDo** Maybe? Restrict some archetypes so a Villain can't pick Bard here either, this mostly just affects the villain as the other absolutely nonsensical choices like the Hero Dread Knight are already blocked by Alignment Restrictions...
**ToDo @JohN100x1 ** Decide if this really should be an Overwhelming feature or would be fairer for the archetypes as a special power

Known Issues:
- Sorcerer and Oracle do not like retroactive selection of a Bloodline or Curse, they are fine with retroactivly applying the features of said Bloodline or Curse, to prevent the game from locking up due to invalid selections this means that if you pick those legacies at higher levels you only get the Bloodlines/Curses/etc. that you would get from that point on and not the past ones. Explanation by example: so if you theoretically select Sorcerer Legacy - Chimera at level 6 you miss out on the level 1 and 3 Bloodline but get the Level 6 one and all future ones and if you pick the arcana bloodline at level 6 you do still get all of its features including the arcane bond selection it had at level 1 at that step.
- Neither the Curse Selection nor the Bloodline Selection count for the basic selection, so if you properly dual class (rather than just select their legacy as a dual class) into those classes later you still have to make those as choices.
- If you use ToyBox to pick multiple classes at each Level you will run into the same issues that you would with Stigmatized Witch and Oracle for your progression calculation, meaning it sums it all up rather than caping it at character level, except since the Protagonist is now patched into most progressions you get it for almost any class that has a progression

**Comment by kjk001** [comment]
I just leveled up in game and saw something I missed
- capstones for the legacies if TT Alternate Capstones are enabled

**Comment by Nerin3** [comment]
"Neither the Curse Selection nor the Bloodline Selection count for the basic selection, so if you properly dual class (rather than just select their legacy as a dual class) into those classes later you still have to make those as choices."
I'd say this one doesn't sound like an issue (If it behaves the way I understand what you wrote).

**Comment by bundillion** [comment]
> "Magus Legacy - Spellblade, Attention Still requires any form of descent testing, so only use it if you are willing to essentially be the betatester for the Legacy at the moment"

Sorry if it's mentioned somewhere else, but is this related to why there isn't a Broad Study arcana option for the Magus to access the isekai class spellbook? I'm interested to learn what the expected issues are for this update, as well.

**Comment by kjk001** [comment]
> "Neither the Curse Selection nor the Bloodline Selection count for the basic selection, so if you properly dual class (rather than just select their legacy as a dual class) into those classes later you still have to make those as choices." I'd say this one doesn't sound like an issue (If it behaves the way I understand what you wrote).

It is an issue because if you multi class witch and oracle the curse  selection of the first will actually count for the second, same if you multi class magus and sorcerer and the bloodline.

So people should be aware that this is not the case if you go Protagonist first and then go for the second class, that means if you only wanted only one curse(for example you are role playing a dhampir from a different basespecies than humand and thus only wanted vampirism) then you should not pick it as a Protagonist feature because the Oracle will force you to pick another.

**Comment by kjk001** [comment]
> > "Magus Legacy - Spellblade, Attention Still requires any form of descent testing, so only use it if you are willing to essentially be the betatester for the Legacy at the moment"
> 
> Sorry if it's mentioned somewhere else, but is this related to why there isn't a Broad Study arcana option for the Magus to access the isekai class spellbook? I'm interested to learn what the expected issues are for this update, as well.

There is no such option because the mod mostly focuses on the protagonist and what to give them. Adding this option for the arcana is possible if people want it. Though there is not much difference for the magus except that you don't have to think about if the spell was in the cleric, wizard, or druid spellist when picking this selection.

Expected issues are:
 - balancing issues amongst the archetypes(given this mod I am not sure this is an actual issue though ;) )
 - things not quite working as expected yet, for example during the last week of playtesting I found that shaman uses yet another mechanic to add the spells granted by its spirits, this method causes a level 0 shaman spellbook to appear on your sheet, ugly but not gamebreaking and any fix I could think of for it would mean we could no longer automatically support shaman spirits added by mods, and I am not sure if there are not some resources still missing that might cause issues in the form of I picked feature x but didn't get the expected bonus y
 - downward compatibility to the previous version, moving features into progressions means that you will have to respec from level 1(unless you are a God Emperor) to rebuild the character.

**Comment by kjk001** [comment]
@JohN100x1 
Updated Changelog and Readme, as for spells see my comment inthe original thread.

When you say update the info.json and repository.json you specifically mean the version number, correct?

So those would be the last things pre release

**Comment by kjk001** [comment]
I also got myself a better tool for merging git repositories, the visual studio internal one is just horrible :D
so future merges should be easier and I expect to be done with the localization file this evening.

**Comment by kjk001** [comment]
or earlier since I just noticed that it had updated that file in my game folder as well so I could just merge it over to the working directory.

I am also now officially rebased on the master branch and would consider TTCMerge closed making new mini branches for all future changes.

**Comment by bundillion** [comment]
> There is no such option because the mod mostly focuses on the protagonist and what to give them. Adding this option for the arcana is possible if people want it. Though there is not much difference for the magus except that you don't have to think about if the spell was in the cleric, wizard, or druid spellist when picking this selection.
That makes sense, I was just thinking about legend builds and throwing spellstrike onto a max-level villain for giggles. As you're already planning the legacy thing, though, I would say giving access via Broad Study isn't all that necessary beyond that particular use case. This is just a question I've had since noticing Broad Study didn't include the protagonist spellbook (i respec'd the levels after realizing the incompatibility, but the legacy would basically sidestep the process of multiclassing in the first place, which is good.)

>     * balancing issues amongst the archetypes(given this mod I am not sure this is an actual issue though ;) )
kermit_sip.png
 
>     * things not quite working as expected yet, ...and I am not sure if there are not some resources still missing that might cause issues in the form of I picked feature x but didn't get the expected bonus y
I sure hope spellstrike doesn't have something weird like this. I just wanna apply Harm and Heal with a scythe! And maybe put out a Polar Ray with a longbow idk.

>     * downward compatibility to the previous version, moving features into progressions means that you will have to respec from level 1(unless you are a God Emperor) to rebuild the character.
I... maybe will hold off on updating to the new system, at least until I finish the playthrough I'm on (or it corrodes under the weight of my poor mod management...)

Thanks for the continued work, y'all. I love this mod, it's great fun.

**Comment by kjk001** [comment]
I will make certain to retest spellstrike specifically throughout the week.

Also, to clarify none of your features would actually be lost in the upgrade, you just wouldn't get additonal sneak attack dice at your following levelups without the respec.

So if you are already at max level the changes also won't mean anything to you.

But yes, finishing whatever run you are on at the moment and then using the new version for a fresh game is a good idea given the scope of the changes.

I would personally actually be interested in how many people switch to a non rogue legacy and what the most popular ones are, as well as feedback on what should still be added to existing templates and what else people would like to see.

**Comment by Mythalar** [comment]
> I... maybe will hold off on updating to the new system, at least until I finish the playthrough I'm on (or it corrodes under the weight of my poor mod management...)
> 

Can't wait to begin a new run with the new version!

**Comment by Mythalar** [comment]
> 
> I would personally actually be interested in how many people switch to a non rogue legacy and what the most popular ones are, as well as feedback on what should still be added to existing templates and what else people would like to see.

Kineticist forever here!
Does the main blast stat will be the one used for casting? 
Would be more complicated if not (Truinto did something for letting us choose which stat we want for psychokineticist it's really great)

Oracle revelations and mysteries will be nice.

I guess for casters (I only play those...) bloodline will be less useful since the progression features are usually less useful than the arcanas themselves (even more for Protagonist regarding immunities , which can be acquired by Isekai feats)

**Comment by kjk001** [comment]
You are patched into the kineticist as normal, so constitution would be the score for that.
Though honestly I would have to check if the hero or edge lord feat to change the damage bonus affect that.

When referencing that someone already did something for a specific archetype in a different mod please always clearly state the mod, preferably with a link, that makes checking on it easier.

Does only the Psychokineticist get that bonus or is it a feat that any kineticist might pick? If it is the second then just loading both mods should grant you that ability already because you count as a full kineticist for the sake of qualifying for feats.

**Comment by kjk001** [comment]
@bundillion 
So Broad Study is only necessary if Tabletop Tweaks Base is installed because that is what disables those features for everything but the magus spells in the first place.

I will update our compatibility patch so that if Broad Study is enabled the magus legacy also gets this Component as a hidden fact rather than as a Selection Option.

**Comment by Mythalar** [comment]
@kjk001 

Sorry for the incomplete info :
[Here](https://github.com/Truinto/DarkCodex/issues/25) you have the original topic on DarCodex repository and the way it was implemented (so not a feat, but a choice in the UMM settings)

**Comment by Mythalar** [comment]
Another point : the size bonus granted by burns would not be useful to any caster Protagonist (or redundant with spell buffs). 

Complicated but would it be a way to select the kineticist archetype, since there is already a choice regarding kinetic knight?

**Comment by Nerin3** [comment]
> I also got myself a better tool for merging git repositories, the visual studio internal one is just horrible :D so future merges should be easier and I expect to be done with the localization file this evening.

Nice, I personally use software called Sourcetree.
Anyways it is good to see that the mod is still worked on.

**Comment by kjk001** [comment]
Undead template should switch any con based check over to charisma anyway, if it does not then that is an error in the template itself that needs patching by owlcat so make sure to send them a bug report.

I can add additional archetypes for separate subtypes, but making it a choice inside the progression would add another level of progression and that might cause issues if we add too many layers. 

So if you have requests for specific subtypes feel free to ask but formulate them as their own legacy archetype and do provide its name and description as I am not very good at that.

Those Archetypes would then also be something for 4.1 as a version and not 4.0. Because if we keep adding to the current build then it will never be released.

@Nerin3 was it ever not? I mean the changes I just made might be rather massive and thus took a few weeks but John has continously worked on bugfixes for the existing features and done review and cleanup on my code.

**Comment by Mythalar** [comment]
Actually the undead template should not change blast main stat to charisma by itself since the burn system is based on non lethal damage, and undeads are immuned to those (that was the point of the Overwhelming Soul archetype).

I'll provide a name and description in a separate topic!

**Comment by kjk001** [comment]
Let me rephrase my statement then, as far as I am aware the game should default to charisma for any con based requirement for undead to prevent bad Nullpointers or Null Divisions. 
Regardless of the fact that some of those features should simply not be possible as an undead based on the more flexible ttrpg ruleset for which such pointers are not causes of a crash but simply open to interpretation.

**Comment by kjk001** [comment]
Anyway here are some screens, bonus points to who can spot what horrible thing the edge lord in the last screen did before the update is released:
![Tactician](https://user-images.githubusercontent.com/121681604/220173716-d381f52d-5396-46d3-9ae2-3e3460c7e6ec.jpg)
![Sorcerer](https://user-images.githubusercontent.com/121681604/220173735-53ca44b4-0cab-4d9b-9a1d-42afefca6613.jpg)
![Rogue](https://user-images.githubusercontent.com/121681604/220173747-d22e19e5-60b8-4dfc-a6be-04a421b4da34.jpg)
![Paladin](https://user-images.githubusercontent.com/121681604/220173762-6cff51c4-763a-46ca-b7aa-a8effff07462.jpg)
![KineticKnight](https://user-images.githubusercontent.com/121681604/220173781-d8b596dc-d56d-4b7a-8ccb-72d7e67a3550.jpg)
Fear my Edge Lord:
![EdgeLord](https://user-images.githubusercontent.com/121681604/220173811-ebda8776-fe47-4dd2-a976-93e9eb234d04.jpg)

and I would say at least for version 4.0.0 I am done

**Comment by Nerin3** [comment]
Now for a trully OP Protagonist class: Legacy Master, every 4 levels they can pick one additional legacy.

**Comment by Mythalar** [comment]
Could be an overwhelming power only selectable at lvl1 :p

**Comment by kjk001** [comment]
and no cookie for Nerin or mythalar

**Comment by Mythalar** [comment]
Horrible thing? Selecting dart as weapon of choice ? xD

**Comment by Nerin3** [comment]
I think it's the Arcane Armor Mastery, since the protagonist doesn't have Arcane Failure "Unless it was added back in."

**Comment by Mythalar** [comment]
> I think it's the Arcane Armor Mastery, since the protagonist doesn't have Arcane Failure "Unless it was added back in."

You are right, I thought it was useless but the raging-darkshadows-edging lord of flying butter knifes would be a truly horrible sight to behold

**Comment by kjk001** [comment]
> > @bundillion So Broad Study is only necessary if Tabletop Tweaks Base is installed because that is what disables those features for everything but the magus spells in the first place.
> 
> ... I feel like I should have been able to figure this out. TT has such a long list of features, too.

either way patch that quietly adds the correct marker to the progression if TTB is enabled has been done and tested.

Arcane Armor Mastery is actually useful if you consider taking another arcane class because you then want to go mystic theurge on them. If not is it useless but not a horrible thing that is cause for fear

**Comment by Mythalar** [comment]
Except the fearsome darts I can only see the 10dex on an archetype based on Dex :p

**Comment by bundillion** [comment]
> either way patch that quietly adds the correct marker to the progression if TTB is enabled has been done and tested.

Actual Platinum-label Rockstar kjk, known icon

**Comment by kjk001** [comment]
paladin abilities rely on charisma even if it is an anti paladin, also everything that was not a new for display feature was simply randomly picked to quickly get to the next point, something I now regret as you focused on everything but the obvious

**Comment by Mythalar** [comment]
Well, except that selecting Ball of Rage seems to have blocked him from selecting nearly half the proposed legacies I can't tell -_-

**Comment by kjk001** [comment]
Edge Lord, Hero, and Villain have their own list of acceptable first choices for their legacy, that is the reason why the list is only half as long and not because of the choice of ball of rage

**Comment by Mythalar** [comment]
Yes I read this but that seemed to be a harsh cut !

But week, might be too obvious for me to find what's so terrifying;)

**Comment by kjk001** [comment]
Not all lists are equally short, I tried to add to every list the choices that make sense for that archetype.
Also check the picked overwhelming feat, you can already pick a second legacy and there are no limits on it besides alignment.
So those apparent fever dreams of you and Nerin are already true.

**Comment by Mythalar** [comment]
Oh I thought I did not see that there was no limits.
An all-party man!

**Comment by Mythalar** [comment]
To be honest it did not strike me  so much that you could select both legacies from the start ^^

**Comment by Nerin3** [comment]
You know what this mod lacks, this:
![I am atomic](https://tenor.com/view/%C4%B1am-atomic-cid-kagenou-shadow-garden-gif-27235465)

**Comment by Magdi76** [comment]
Hello, please tell me - what is TTCCore?

**Comment by bundillion** [comment]
> Hello, please tell me - what is TTCCore?

"Tabletop Tweaks" is a 3-part mod that alters a bunch (like a holy hell of a tonne) of rules within the game so it works closer to the PF1e game. It was made in response to Owlcat truncating a whole bunch of mechanics from the core system in order to make the videogame part of WotR (and Kingmaker) a bit more videogame-like rather than ttrpg-like.

(one of those options adds the Broad Study magus arcana, which is what caused the issue I brought up in the first place lol)

**Comment by JohN100x1** [comment]
I think we need a standardised .csproj file because everytime I try to build I have to change it back to my setup

**Comment by bundillion** [comment]
I'm puking blood, the 2.10u update broke scaling cantrips and, therefore, my progress. Guess I'll just wait for this to update for the next run. Gonna go get a mop for all this red bile.

**Comment by JohN100x1** [comment]
> * patch for some other mod that turned deity selection into a recursive selection call onto itself which made it hard to select deities

@kjk001 Which mod is this exactly? The new code seems to give me a null reference exception when I try to load the game
```
NullReferenceException: Object reference not set to an instance of an object
  at Kingmaker.TextTools.NameTemplate.Generate (System.Boolean capitalized, System.Collections.Generic.List`1[T] parameters) [0x00027] in <7233d069edc947ad9245aca7f2d9f2b6>:0 
  at Kingmaker.TextTools.TextTemplateEngine.Process (System.String text) [0x00242] in <7233d069edc947ad9245aca7f2d9f2b6>:0 
  at Kingmaker.Localization.LocalizedString.op_Implicit (Kingmaker.Localization.LocalizedString localizedString) [0x00069] in <7233d069edc947ad9245aca7f2d9f2b6>:0 
  at Kingmaker.Blueprints.Facts.BlueprintUnitFact.get_Name () [0x00000] in <7233d069edc947ad9245aca7f2d9f2b6>:0 
  at TabletopTweaks.Core.Utilities.BlueprintExtentions+<>c.<AddFeatures>b__18_0 (Kingmaker.Blueprints.BlueprintFeatureReference feature) [0x00010] in <74e6287844a240d1a54a73c18234bf69>:0 
  at System.Linq.EnumerableSorter`2[TElement,TKey].ComputeKeys (TElement[] elements, System.Int32 count) [0x00010] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at System.Linq.EnumerableSorter`1[TElement].ComputeMap (TElement[] elements, System.Int32 count) [0x00000] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at System.Linq.EnumerableSorter`1[TElement].Sort (TElement[] elements, System.Int32 count) [0x00000] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at System.Linq.OrderedEnumerable`1[TElement].SortedMap (System.Linq.Buffer`1[TElement] buffer) [0x00006] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at System.Linq.OrderedEnumerable`1[TElement].ToArray () [0x00025] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at System.Linq.Enumerable.ToArray[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x0001f] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at TabletopTweaks.Core.Utilities.BlueprintExtentions.AddFeatures (Kingmaker.Blueprints.Classes.Selection.BlueprintFeatureSelection selection, Kingmaker.Blueprints.BlueprintFeatureReference[] features) [0x0007f] in <74e6287844a240d1a54a73c18234bf69>:0 
  at TabletopTweaks.Core.Utilities.BlueprintExtentions.AddFeatures (Kingmaker.Blueprints.Classes.Selection.BlueprintFeatureSelection selection, Kingmaker.Blueprints.Classes.BlueprintFeature[] features) [0x0002b] in <74e6287844a240d1a54a73c18234bf69>:0 
  at IsekaiMod.Content.Deities.IsekaiDeitySelection.Add () [0x0002f] in <9a50b11055804b4097dff00567a788dc>:0 
  at IsekaiMod.Content.ContentAdder+BlueprintsCache_Init_Patch.AddIsekaiDeities () [0x00000] in <9a50b11055804b4097dff00567a788dc>:0 
  at IsekaiMod.Content.ContentAdder+BlueprintsCache_Init_Patch.Postfix () [0x00073] in <9a50b11055804b4097dff00567a788dc>:0 
  at (wrapper dynamic-method) Kingmaker.Blueprints.JsonSystem.BlueprintsCache.Kingmaker.Blueprints.JsonSystem.BlueprintsCache.Init_Patch5(Kingmaker.Blueprints.JsonSystem.BlueprintsCache)
  at (wrapper dynamic-method) Kingmaker.Blueprints.JsonSystem.StartGameLoader.Kingmaker.Blueprints.JsonSystem.StartGameLoader.LoadPackTOC_Patch2(Kingmaker.Blueprints.JsonSystem.StartGameLoader)
  at Kingmaker.GameStarter+<StartGameCoroutine>d__25.MoveNext () [0x00375] in <7233d069edc947ad9245aca7f2d9f2b6>:0 
  at UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) [0x00026] in <25a85da7c6f04932b86e339dfd12957d>:0 

```

**Comment by kjk001** [comment]
I don't know what mod did it, but that place you changed back also has nothing to do with the actual fix for that recursive nightmare, because the patch function below is called much later than the adding of the deities.

If the sorting there crashes for you due to a null name then likely at least one mod in your collection also screws with deities in a way it shouldn't.

What is a bit weird is that I know that at least one of my mods also does this because I get a null value warning when I do the sorting at the end but for me the ttc way does not cause a crash.

Shelving and unshelving the csproj file with no further changes should work shouldn't it?

**Comment by JohN100x1** [comment]
Ok the release has just been published. If there's any bugs that occur we add it to the 4.1.0 checklist (#81) to deal with.
I also created a [new branch](https://github.com/JohN100x1/WOTR_IsekaiMod/tree/version-4-1-0) so anything being implemented should go in there rather than the master branch. That way I can see all the changes before merging to master.
